### DISCUSSION TITLE
Is it possible to run multiple bot on one project?

### DISCUSSION THREAD

**Discussion by 100nandoo** [aiogram/aiogram#578] [Q&A]
If it is possible, can provide an example

**Comment by Olegt0rr** [comment]
> If it is possible, can provide an example

```python
tokens = [
    first_token, 
    second_token,
    third_token,
]
bot_tasks = []
for token in tokens:
    bot = Bot(token)
    dp = Dispatcher(bot)
    bot_task = asyncio.create_task(dp.start_polling())
    bot_tasks.append(bot_task)

await asyncio.gather(*bot_tasks)

```

**Reply by 100nandoo** [reply]
I tried it just now, and it still doesn't work. only the first bot is active

**Reply by 100nandoo** [reply]
```py
async def start() -> None:
    logger.add(sys.stderr, format="{time:YYYY-MM-DD at HH:mm:ss} | {level} | {message}")
    logger.info(f"start polling for {__name__}")
    loop = asyncio.get_event_loop()
    loop.run_until_complete(await captain_dp.start_polling())
    loop.run_until_complete(await dp.start_polling())
```

**Reply by evgfilim1** [reply]
Try this: (just for your case)

```python
async def start() -> None:
    logger.add(sys.stderr, format="{time:YYYY-MM-DD at HH:mm:ss} | {level} | {message}")
    logger.info(f"start polling for {__name__}")
    await asyncio.gather(captain_dp.start_polling(), dp.start_polling())
```

**Reply by 100nandoo** [reply]
Thank you very much, it's working!
I can publish an example if anybody interested to run multiple bot in one heroku instance (to save some dyno hours).

**Reply by evgfilim1** [reply]
> I can publish an example if anybody interested to run multiple bot in one heroku instance (to save some dyno hours).

Feel free to do this in [Show and tell](https://github.com/aiogram/aiogram/discussions/categories/show-and-tell) section of our Discussions. Maybe someone will find this useful :wink:

**Reply by 100nandoo** [reply]
Alright I will post it soon.

**Reply by maksymskuibida** [reply]
Give an example, please, I need to start up to thousands bots. Every bot eats 100-200mb memory and now with 40+ bots it eats more memory then we have on server

**Reply by maksymskuibida** [reply]
I tried to do it, but states don`t work((

**Reply by maksymskuibida** [reply]
tried*

**Reply by Olegt0rr** [reply]
> Give an example, please, I need to start up to thousands bots. Every bot eats 100-200mb memory and now with 40+ bots it eats more memory then we have on server

It's a bad idea to use polling for that case. Use webhooks

**Comment by humza-sami** [comment]
I know its quite late but i fixed it like this:

   ```python
 import asyncio

    async def start_raid_bot():
        dp_bot1 = Dispatcher()
        dp_bot1.include_router(raid_router)
        await dp_bot1.start_polling(bot1)
    
    async def start_trend_bot():
        dp_bot2 = Dispatcher()
        dp_bot2.include_router(router_trend_bot)
        await dp_bot2.start_polling(bot2)
    
    async def main():
        asyncio.create_task(confirm_payment_and_inform_user())
        task_bot1 = asyncio.create_task(start_raid_bot())
        task_bot2 = asyncio.create_task(start_trend_bot())
        await task_bot1
        await task_bot2
    
    if __name__ == "__main__":
        asyncio.run(main())


```
